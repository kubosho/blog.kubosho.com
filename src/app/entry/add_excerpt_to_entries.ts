import { convertMarkdownToPlainText } from './entry_converter';
import type { TinyCollectionEntry } from './tiny_collection_entry';

function extractFirstNonEmptyLine(body: string): string {
  return (
    body
      .split('\n')
      .filter((b) => b !== '')
      .at(0) ?? ''
  );
}

export async function addExcerptToEntries(entries: TinyCollectionEntry[]): Promise<TinyCollectionEntry[]> {
  return await Promise.all(
    entries.map(async (entry) => {
      const hasExcerpt = entry.data.excerpt != null;
      if (hasExcerpt) {
        return entry;
      }

      const excerpt = (await convertMarkdownToPlainText(extractFirstNonEmptyLine(entry.body))).trim();

      return {
        ...entry,
        data: {
          ...entry.data,
          excerpt,
        },
      };
    }),
  );
}
