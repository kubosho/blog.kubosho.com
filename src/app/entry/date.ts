import dayjs from 'dayjs';
import timezone from 'dayjs/plugin/timezone';
import utc from 'dayjs/plugin/utc';

dayjs.extend(utc);
dayjs.extend(timezone);

export function convertIsoStringToMilliseconds(dateIsoString: string): number {
  const r = dayjs(dateIsoString).unix() * 1000;
  return r;
}

export function formatRfc2822(dateTimeMilliseconds: number): string {
  const utc = dayjs(dateTimeMilliseconds).utc();
  const r = utc.format('ddd, DD MMM YYYY HH:mm:ss ZZ');
  return r;
}

export function formatIsoString(dateTimeMilliseconds: number): string {
  const utc = dayjs(dateTimeMilliseconds).utc();
  const r = utc.toISOString();
  return r;
}

export function formatYYMMDDString(dateTimeMilliseconds: number, separator = '.'): string {
  const utc = dayjs(dateTimeMilliseconds).utc();
  const r = utc.format(`YYYY${separator}MM${separator}DD`);
  return r;
}
