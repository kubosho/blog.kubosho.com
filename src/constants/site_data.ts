export const SITE_DESCRIPTION =
  '学習し、自分の中で落とし込んだあとに、言語化をして書き出すブログ。主な話題はソフトウェア開発ですが、日記やゲーム・ガジェット・ダーツなど自分の趣味の話題も書きます。';
export const SITE_TITLE = '学ぶ、考える、書き出す。';

export const AUTHOR = 'kubosho';
export const AUTHOR_WITH_UNDERSCORE = `${AUTHOR}_`;

export const SITE_URL = 'https://blog.kubosho.com';
export const LOCAL_SITE_URL = 'http://localhost:3000';

export const TWITTER_ACCOUNT_ID = AUTHOR_WITH_UNDERSCORE;
export const FACEBOOK_APP_ID = '2453282784920956';
