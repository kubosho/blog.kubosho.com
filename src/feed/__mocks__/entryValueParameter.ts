import { EntryValueParameter } from '../../entry/entryValue';

export const mockEntryValueParameter: EntryValueParameter = {
  body:
    '<p>2019 年は丸 1 年ブログの記事を書かず、2018 年も長い記事を書いていなかった。\nそのせいか書きたい題材はあるものの、記事の内容が思いつかない、もしくは書こうとしても面倒くさくなるということになっている。</p>\n<p>とはいえ自分の Twitter のタイムラインを見ていると、さまざまな人がブログ・ポッドキャスト・YouTube などといったさまざまな手段で発信している。</p>\n<p>中にはサービスを作って公開している人もいる。</p>\n<p>そういうのを見ると、自分の中に羨ましい・楽しそうという気持ちが芽生えて、ああこの流れに自分も乗っていきたいという気持ちが高まる。\nそうしてまずはブログの記事から書こうと思うものの、内容が思いつかない、または書こうとしても面倒くさくなるということが起きる。</p>\n<p>まあこういうのもさらけ出して、自分の中の発信のハードルを下げていって量を出せるようにしていけば、いつか質も伴った発信ができるようになるのかなと思う。\nブログだけでなくせっかく機材が揃っているのでポッドキャストや YouTube などを使った発信というのもしていきたいな。</p>\n',
  excerpt: '2019 年は丸 1 年ブログの記事を書かず、2018 年も長い記事を書いていなかった。\n',
  id: 'content-of-article-can-not-think-of-anything-syndrome',
  slug: 'content-of-article-can-not-think-of-anything-syndrome',
  title: '記事の内容が思いつかない症候群',
  categories: ['日記'],
  tags: ['考え事'],
  createdAt: new Date('2020-05-25').toISOString(),
  updatedAt: new Date('2020-05-25').toISOString(),
  publishedAt: new Date('2020-05-25').toISOString(),
  revisedAt: new Date('2020-05-25').toISOString(),
};
