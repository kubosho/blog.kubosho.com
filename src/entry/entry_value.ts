import { unwrapMaybe } from 'option-t/lib/Maybe/unwrap';
import { BlogApiSchema } from '../microcms_api/api_schema';
import { HeroImage } from '../microcms_api/hero_image';

import { convertISOStringToMilliseconds } from './date';

export class EntryValue {
  readonly id: string;
  readonly slug: string;
  readonly title: string;
  readonly body: string;
  readonly publishedAt: number;
  readonly revisedAt: number;
  readonly excerpt?: string;
  readonly heroImage?: HeroImage;
  readonly categories?: string[];
  readonly tags?: string[];

  constructor(param: BlogApiSchema) {
    const c = unwrapMaybe(param.publishedAt || param.createdAt);
    const u = unwrapMaybe(param.revisedAt || param.updatedAt);

    const publishedAt = convertISOStringToMilliseconds(c);
    const revisedAt = convertISOStringToMilliseconds(u);

    this.id = unwrapMaybe(param.id);
    this.slug = unwrapMaybe(param.slug || param.id);
    this.title = unwrapMaybe(param.title);
    this.body = unwrapMaybe(param.body);
    this.excerpt = param.excerpt ?? null;
    this.heroImage = param.heroImage ?? null;
    this.categories = param.categories ?? null;
    this.tags = param.tags ?? null;
    this.publishedAt = publishedAt;
    this.revisedAt = revisedAt;
  }
}
