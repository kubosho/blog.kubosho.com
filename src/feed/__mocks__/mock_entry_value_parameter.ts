import { EntryValueParameter } from '../../entry/entry_value';

export const mockEntryValueParameter: EntryValueParameter = {
  body: '<p>2019 年は丸 1 年ブログの記事を書かず、2018 年も長い記事を書いていなかった。</p>',
  excerpt: '2019 年は丸 1 年ブログの記事を書かず、2018 年も長い記事を書いていなかった。\n',
  id: 'content-of-article-can-not-think-of-anything-syndrome',
  slug: 'content-of-article-can-not-think-of-anything-syndrome',
  title: '記事の内容が思いつかない症候群',
  categories: ['日記'],
  tags: ['考え事'],
  heroImage: undefined,
  createdAt: new Date('2020-05-25').toISOString(),
  updatedAt: new Date('2020-05-25').toISOString(),
  publishedAt: new Date('2020-05-25').toISOString(),
  revisedAt: new Date('2020-05-25').toISOString(),
};
