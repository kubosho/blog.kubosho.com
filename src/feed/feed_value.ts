import { pathList } from '../../constants/path_list';
import { EntryValue } from '../entry/entry_value';

export interface WebSiteMetadata {
  title: string;
  baseUrl: string;
  buildTime: string;
}

export interface FeedValue {
  metadata: WebSiteMetadata;
  items: {
    title: string;
    content: string;
    summary: string;
    link: string;
    published: number;
    updated: number;
  }[];
}

export function createFeedValue(entries: readonly EntryValue[], metadata: WebSiteMetadata): FeedValue {
  const items = entries.map((entry) => {
    const link = `${metadata.baseUrl}${pathList.entries}/${entry.slug}`;
    const published = entry.publishedAt;
    const updated = entry.revisedAt;

    return {
      title: entry.title,
      content: entry.body,
      summary: entry.excerpt,
      published,
      updated,
      link,
    };
  });

  return {
    metadata,
    items,
  };
}
