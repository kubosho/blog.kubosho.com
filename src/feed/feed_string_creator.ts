import escapeHtml from 'escape-html';

import { pathList } from '../constants/path_list';
import { AUTHOR, BASE_LANGUAGE } from '../constants/site_data';
import { formatISOString, formatYYMMDDString } from '../entry/date';

import { FeedValue } from './feed_value';

export type XmlString = string;

export function createXmlString(feedValue: FeedValue, baseUrl: string): XmlString {
  return `<?xml version="1.0" encoding="utf-8" ?>
<feed xmlns="http://www.w3.org/2005/Atom" xml:lang="${BASE_LANGUAGE}">
${createMetaXmlString(feedValue, baseUrl)}
${createItemsXmlString(feedValue)}
</feed>`;
}

function createMetaXmlString(feedValue: FeedValue, baseUrl: string): XmlString {
  const { title, link } = feedValue.channel;
  const { host } = new URL(baseUrl);

  return `<title>${title}</title>
<id>tag:${host},2014:feed</id>
<author>
  <name>${AUTHOR}</name>
</author>
<updated>${process.env.BUILD_TIME}</updated>
<link rel="alternate" href="${link}"/>
<link rel="self" type="application/atom+xml" href="${baseUrl}${pathList.feed}"/>`;
}

function createItemsXmlString(feedValue: FeedValue): XmlString {
  const xmlStrings = feedValue.items.map((item) => {
    const { host, pathname } = new URL(item.link);
    const itemId = pathname.slice(pathname.lastIndexOf('/'));

    return `<entry>
<title>${item.title}</title>
<link href="${item.link}" rel="alternate"/>
<id>tag:${host},${formatYYMMDDString(item.published, '-')}:entry:/${itemId}</id>
<content type="html">${escapeHtml(item.content)}</content>
<summary>${item.summary}</summary>
<published>${formatISOString(item.published)}</published>
<updated>${formatISOString(item.updated)}</updated>
</entry>`;
  });

  return xmlStrings.join('\n');
}
