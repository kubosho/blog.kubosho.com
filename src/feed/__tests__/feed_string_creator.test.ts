import { expect, test } from 'vitest';

import { SITE_URL } from '../../../constants/site_data';
import { EntryValue } from '../../entry/entry_value';
import { mockEntryValueParam } from '../__mocks__/mock_entry_value_parameter';
import { createXmlString } from '../feed_string_creator';
import { createFeedValue } from '../feed_value';

test('Feed correctly', () => {
  const metadata = {
    title: 'test',
    baseUrl: SITE_URL,
    buildTime: '2023-02-08T08:10:03.979Z',
  };
  const entryValue = new EntryValue(mockEntryValueParam);
  const feedValue = createFeedValue([entryValue], metadata);
  const tree = createXmlString(feedValue);
  expect(tree).toMatchSnapshot();
});
