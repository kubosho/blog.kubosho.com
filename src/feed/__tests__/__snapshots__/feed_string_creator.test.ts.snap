// Vitest Snapshot v1

exports[`Feed correctly 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\" ?>
<feed xmlns=\\"http://www.w3.org/2005/Atom\\" xml:lang=\\"ja\\">
<title>test</title>
<id>tag:blog.kubosho.com,2014:feed</id>
<author>
  <name>kubosho</name>
</author>
<updated>undefined</updated>
<link rel=\\"alternate\\" href=\\"https://blog.kubosho.com\\"/>
<link rel=\\"self\\" type=\\"application/atom+xml\\" href=\\"https://blog.kubosho.com/feed\\"/>
<entry>
<title>記事の内容が思いつかない症候群</title>
<link href=\\"https://blog.kubosho.com/entries/content-of-article-can-not-think-of-anything-syndrome\\" rel=\\"alternate\\"/>
<id>tag:blog.kubosho.com,2020-05-25:entry://content-of-article-can-not-think-of-anything-syndrome</id>
<content type=\\"html\\">&lt;p&gt;2019 年は丸 1 年ブログの記事を書かず、2018 年も長い記事を書いていなかった。&lt;/p&gt;</content>
<summary>2019 年は丸 1 年ブログの記事を書かず、2018 年も長い記事を書いていなかった。
</summary>
<published>2020-05-25T00:00:00.000Z</published>
<updated>2020-05-25T00:00:00.000Z</updated>
</entry>
</feed>"
`;
